### PR TITLE
Add SyncOptions fallback

### LINKED ISSUE #82: Add fallback  default on SyncOptions
## Description

Still always `Option<SyncOptions>`. On creation of account we either set a `SyncOptions` or we set `SyncOptions::default()`. when passing None, we use that one we set instead of always calling `SyncOptions::default`

## Motivation

So we dont need to manage our SyncOptions ourselves but can set and forget unless we need something specific

### PR DESCRIPTION
**Pull request by kwek20**
# Description of change

- [x] Add as an accountdetails method
- [x] Leave the option to provide custom syncoptions
- [x] Persist to storage
- [x] Add to bindings? (excl java)

## Links to any relevant issues

fixes https://github.com/iotaledger/iota-sdk/issues/82

## Type of change

- Enhancement (a non-breaking change which adds functionality)

### CODE REVIEW

File: sdk/src/wallet/account/operations/syncing/mod.rs [hunk 1]
```diff
@@ -24,10 +24,24 @@ use crate::{
 };
 
 impl AccountHandle {
+    /// Set the fallback SyncOptions for account syncing
+    pub async fn set_default_sync_options(&self, options: SyncOptions) {
+        *self.fallback_sync_options.lock().await = options;
+    }
+
+    // Get the default sync options we use when none are provided
+    pub async fn sync_options(&self) -> SyncOptions {
```
Comment (Thoralf-M) [inline sdk/src/wallet/account/operations/syncing/mod.rs]: Name it `default_sync_options()` so it's aligned?

Comment (kwek20) [reply sdk/src/wallet/account/operations/syncing/mod.rs]: i thought it was ugly :D i guess we like alignment though :(

File: sdk/src/wallet/account/handle.rs [hunk 2]
```diff
@@ -46,6 +46,7 @@ pub struct FilterOptions {
 #[derive(Debug, Clone)]
 pub struct AccountHandle {
     account: Arc<RwLock<Account>>,
+    pub(crate) fallback_sync_options: Arc<Mutex<SyncOptions>>,
```
Comment (Thoralf-M) [inline sdk/src/wallet/account/handle.rs]: This needs to be set on every new start of an application, shouldn't we also persist it?

File: sdk/src/wallet/storage/manager.rs [hunk 3]
```diff
@@ -181,4 +181,18 @@ impl StorageManager {
             .set(ACCOUNTS_INDEXATION_KEY, self.account_indexes.clone())
             .await
     }
+
+    pub async fn set_default_sync_options(
+        &mut self,
+        account_index: u32,
+        sync_options: &SyncOptions,
+    ) -> crate::wallet::Result<()> {
+        let key = format!("{ACCOUNT_INDEXATION_KEY}{account_index}-{ACCOUNT_SYNC_OPTIONS}");
```
Comment (kwek20) [inline sdk/src/wallet/storage/manager.rs:190]: opinions on this key?

Comment (Thoralf-M) [reply sdk/src/wallet/storage/manager.rs:190]: Why not just "{ACCOUNT_SYNC_OPTIONS}{account_index}"?

File: sdk/src/wallet/account/operations/syncing/mod.rs [hunk 4]
```diff
@@ -23,10 +23,33 @@ use crate::{
...
+    }
+
     /// Sync the account by fetching new information from the nodes. Will also retry pending transactions
-    /// if necessary.
+    /// if necessary. A custom default can be set using set_default_sync_options.
     pub async fn sync(&self, options: Option<SyncOptions>) -> crate::wallet::Result<AccountBalance> {
-        let options = options.unwrap_or_default();
+        let options = match options {
+            Some(opt) => opt,
+            None => self.default_sync_options().await.clone(),
```
Comment (Thoralf-M) [inline sdk/src/wallet/account/operations/syncing/mod.rs]: This clone shouldn't be needed? Or you remote the clone in `default_sync_options()` to return `&SyncOptions`

Comment (kwek20) [reply sdk/src/wallet/account/operations/syncing/mod.rs]: removed the clone, couldnt make the ref return as its in a MutexGuard

File: sdk/src/wallet/storage/manager.rs (hunk 3 above)
Comment (kwek20) [reply sdk/src/wallet/storage/manager.rs:190]: I prefer keeping the same format as setting an account, as this is on top of account, not apart

File: sdk/src/wallet/account/handle.rs (hunk 2 above)
Comment (kwek20) [reply sdk/src/wallet/account/handle.rs]: done

File: sdk/src/wallet/account/operations/syncing/mod.rs (hunk 1 above)
Comment (kwek20) [reply sdk/src/wallet/account/operations/syncing/mod.rs]: Done

File: sdk/src/wallet/storage/manager.rs (hunk 3 above)
Comment (Thoralf-M) [reply sdk/src/wallet/storage/manager.rs:190]: I'm also fine with it

File: bindings/python/iota_sdk/wallet/account.py [hunk 5]
```diff
@@ -448,6 +449,16 @@ def set_alias(self, alias):
             }
         )
 
+    def set_default_sync_options(self, options):
+        """Set the fallback SyncOptions for account syncing.
+           If storage is enabled, will persist during restarts.
+        """
+        return self._call_account_method(
+            'SetDefaultSyncOptions', {
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/wallet/account.py]:
```suggestion
            'setDefaultSyncOptions', {
```

**Review by Thoralf-M — DISMISSED** [review]

File: sdk/CHANGELOG.md [hunk 6]
```diff
@@ -64,6 +65,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Fields of `AccountBalance`, `BaseCoinBalance` and `NativeTokensBalance` have been made private and getters have been added;
 - Exposed `AccountParticipationOverview, ParticipationEventWithNodes, AliasOutputOptions, AliasOutputOptionsDto, IncreaseNativeTokenSupplyOptions, IncreaseNativeTokenSupplyOptionsDto, NativeTokenOptions, NativeTokenOptionsDto, NftOptions, NftOptionsDto, OutputOptionsDto` from the `account` module;
 - Made `Wallet::get_bech32_hrp()` public;
+- Made `Account::new()` `async` and return `Result<Self>`;
```
Comment (thibault-martinez) [inline sdk/CHANGELOG.md]: Not really necessary since this is not public

File: sdk/CHANGELOG.md [hunk 7]
```diff
@@ -50,6 +50,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `{BaseCoinBalance, BaseCoinBalanceDto}::voting_power` field;
 - `verify_mnemonic()`;
 - `SecretManager::sign_transaction()`;
+- `Account::{set_default_sync_options, default_sync_options}` methods;
```
Comment (thibault-martinez) [inline sdk/CHANGELOG.md]: You're adding to 0.2, not 0.3

File: sdk/src/wallet/message_interface/account_method.rs [hunk 8]
```diff
@@ -288,7 +288,7 @@ pub enum AccountMethod {
         max_attempts: Option<u64>,
     },
     /// Sync the account by fetching new information from the nodes. Will also retry pending transactions
-    /// if necessary.
+    /// if necessary. A custom default can be set using SetDefaultSyncOptions
```
Comment (thibault-martinez) [inline sdk/src/wallet/message_interface/account_method.rs]:
```suggestion
    /// if necessary. A custom default can be set using SetDefaultSyncOptions.
```

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: sdk/src/wallet/account/operations/syncing/mod.rs [hunk 9]
```diff
@@ -23,10 +23,33 @@ use crate::{
...
+            let index = *self.read().await.index();
+            let mut storage_manager = self.storage_manager.lock().await;
+            storage_manager.set_default_sync_options(index, &options).await?;
+        }
+
+        *self.default_sync_options.lock().await = options;
+        Ok(())
+    }
+
+    // Get the default sync options we use when none are provided
```
Comment (thibault-martinez) [inline sdk/src/wallet/account/operations/syncing/mod.rs]:
```suggestion
    // Get the default sync options we use when none are provided.
```

**Review by thibault-martinez — APPROVED** [review]

**Review by Thoralf-M — APPROVED** [review]

**Pull request merged**